src/Nixfmt/Predoc.hs

```diff
@@ -55,32 +48,31 @@ data Spacing
     | Newlines Int
     deriving (Show, Eq, Ord)
 
-data Predoc f
+data DocAnn
+    -- | Node Group docs indicates either all or none of the Spaces and Breaks
+    -- in docs should be converted to line breaks.
+    = Group
+    -- | Node (Nest n) docs indicates all line start in docs should be indented
+    -- by n more spaces than the surroundings.
+    | Nest Int
+    deriving (Show, Eq)
+
+data DocE
     = Text Text
     | Spacing Spacing
-    -- | Group predoc indicates either all or none of the Spaces and Breaks in
-    -- predoc should be converted to line breaks.
-    | Group (f (Predoc f))
-    -- | Nest n predoc indicates all line start in predoc should be indented by
-    -- n more spaces than the surroundings.
-    | Nest Int (f (Predoc f))
+    | Node DocAnn Doc
+    deriving (Show, Eq)
 
-deriving instance Eq (Predoc Tree)
-deriving instance Eq (Predoc [])
-deriving instance Show (Predoc Tree)
-deriving instance Show (Predoc [])
-
-type Doc = Tree (Predoc Tree)
-type DocList = [Predoc []]
+type Doc = [DocE]
```

Comment:
I'm not sure how much files this tool is expected to work with, and probably this concern has already been raised, but have you considered replacing `[]` with something else? E.g. on each nested code block you perform concatenation of documentation items and it takes linear time.

Usually, for an incremental building they use data structures which provide lazy concatenation with `O(1)` amortized complexity. For instance, `DList` can be picked; although, it would be slightly less convenient to use.

Reply:
Actually, I used a custom data structure before as this diff shows and it made the code significantly more complicated. From what I can tell, using lists is less than 20% slower on `nixpkgs`, so this will do for now. I can look at optimization in the future, but I want it to be guided by profiling.